src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs

```diff
@@ -179,16 +179,121 @@ public static void ToApmEnd(this IAsyncResult iar)
             }
         }
 
-        // Task.GetAwaiter().GetResult() calls an internal variant of Wait() which doesn't wrap exceptions in
-        // an AggregateException.
-        public static void WaitForCompletion(this Task task)
+        public static void RunSyncWithParams<T1>(this Func<T1, Task> function, T1 param1)
         {
-            task.GetAwaiter().GetResult();
+            var state = Tuple.Create(function, param1);
+            var syncTask = new Task<Task>(obj => {
+                var tuple = obj as Tuple<Func<T1, Task>, T1>;
+                var func = tuple.Item1;
+                var p1 = tuple.Item2;
+                return func(p1);
+            }, state, CancellationToken.None);
```

Comment:
There are two timeouts at play here. The function being passed in handles any timeouts or cancellations. For example RequestChannel.RequestAsyncInternal has the following method signature:
```csharp
private async Task<Message> RequestAsyncInternal(Message message, TimeSpan timeout)
```
This has two parameters, message and timeout, which will map to param1 and param2. Inside the delegate, the function us passed these two parameters. If one of them was a cancellation token, it would get passed in too. It's up to the function to handle the cancellation token or timeout that is passed in. In this particular overload, a new task object is instantiated beginning on line 185. It is this task which has a CancellationToken.None attached to it and not the inner function being called. The significance of this is that waiting on this outer Task means there is no concern about needing to stop waiting for it to complete before the actual execution has completed. E.g. if the inner function took 10 seconds to complete and the outer Task was constructed with a CancellationToken which cancels after 5 seconds, then the code couldn't be executed on the same thread as the method that is waiting on the Task would need to be returned to once the CT has cancelled which would happen after 5 seconds. In this case, the execution actually happens on a different thread so that the current thread is available in the case of a potential cancellation.